public/modules/custom/helfi_ahjo/src/Services/AhjoService.php

```diff
@@ -0,0 +1,292 @@
+    return new static(
+      $container->get('extension.list.module'),
+      $container->get('helfi_ahjo.taxonomy_utils'),
+      $container->get('entity_type.manager'),
+      $container->get('http_client'),
+      $container->get('messenger')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getConfig(): ImmutableConfig {
+    return \Drupal::config('helfi_ahjo.config');
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function fetchDataFromRemote($orgId = 00001, $maxDepth = 9999): array {
+    $config = self::getConfig();
+    if (strlen($orgId) < 5) {
+      $orgId = sprintf('%05d', $orgId);
+    }
+
+    if (strlen($maxDepth) < 4) {
+      $maxDepth = sprintf('%04d', $maxDepth);
+    }
+
+    $url = sprintf("%s/fi/ahjo-proxy/org-chart/$orgId/$maxDepth?api-key=%s", $config->get('base_url'), $config->get('api_key'));
+
+    $response = $this->guzzleClient->request('GET', $url);
+    return Json::decode($response->getBody()->getContents());
+  }
+
+  /**
+   * @param array $childData
+   * @param array $operations
+   * @param int $externalParentId
+   *
+   * Recursive set all information from ahjo api.
+   *
+   */
+  public function setAllBatchOperations($childData = [], &$operations = [], $externalParentId = 0): void {
+    foreach ($childData as $content) {
+      $content['externalParentId'] = $externalParentId;
+      $operations[] = ['_helfi_ahjo_batch_dispatcher',
+        [
+          'helfi_ahjo.ahjo_service:syncTaxonomyTermsOperation',
+          $content,
+        ],
+      ];
+
+      if (isset($content['OrganizationLevelBelow'])) {
+        $this->setAllBatchOperations($content['OrganizationLevelBelow'], $operations, $content['ID']);
+      }
+    }
+
+  }
+
+  /**
+   * Create batch operations for taxonomy sote_section.
+   *
+   * @param array $data
+   *   Data for batch.
+   *
+   * @return void
+   */
+  public function createTaxonomyBatch(array $data) {
+    $operations = [];
+
+    $this->setAllBatchOperations($data, $operations);
+
+    $batch = [
+      'operations' => $operations,
+      'finished' => [AhjoService::class, 'syncTermsBatchFinished'],
+      'title' => 'Performing an operation',
+      'init_message' => 'Please wait',
+      'progress_message' => 'Completed @current from @total',
+      'error_message' => 'An error occurred',
+    ];
+
+    batch_set($batch);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function addToCron($data, $queue, $parentId = NULL) {
+    if (!is_array($data)) {
+      $data = Json::decode($data);
+    }
+
+    foreach ($data as $section) {
+      $section['externalParentId'] = $parentId ?? 0;
+      $queue->createItem($section);
+
+      if (isset($section['OrganizationLevelBelow'])) {
+        $this->addToCron($section['OrganizationLevelBelow'], $queue, $section['ID']);
+      }
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function syncTaxonomyTermsChilds() {
+//    $terms = $this->entityTypeManager
```

Comment:
if the sync tree works ine with the changes that we did on parent id, let's remove this